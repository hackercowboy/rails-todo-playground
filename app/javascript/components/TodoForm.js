import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Formik, Field, Form } from 'formik';

function TodoForm({ onSubmit, onError }) {
  const [fetching, setFetching] = useState(false);
  const input = React.useRef();

  const handleOnSubmit = async (values, actions) => {
    setFetching(true);
    try {
      await onSubmit(values);
      actions.resetForm({ text: '' });
    } catch (error) {
      onError(error.response.data.errors);
    } finally {
      setFetching(false);
      if (input.current) {
        input.current.focus();
      }
    }
  };

  useEffect(() => {
    if (input.current) {
      input.current.focus();
    }
  });

  return (
    <Formik initialValues={{ text: '' }} onSubmit={handleOnSubmit}>
      <Form className="todo-form">
        <Field
          id="text"
          name="text"
          type="text"
          placeholder="Add new Todo"
          disabled={fetching}
          innerRef={input}
        />
      </Form>
    </Formik>);
}

TodoForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  onError: PropTypes.func.isRequired,
};

export default TodoForm;
