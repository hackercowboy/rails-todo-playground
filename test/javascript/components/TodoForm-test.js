import React from 'react';
import { render, prettyDOM } from '@testing-library/react';

import TodoForm from '../../../app/javascript/components/TodoForm';

describe('<TodoForm/>', () => {
  it('should render', () => {
    render(<TodoForm onSubmit={jest.fn()} onError={jest.fn()} />);
    expect(prettyDOM()).toMatchSnapshot();
  });
});
