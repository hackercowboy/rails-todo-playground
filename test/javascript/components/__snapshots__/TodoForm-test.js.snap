// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoForm/> should render 1`] = `
"<Formik initialValues={{...}} onSubmit={[Function: handleOnSubmit]}>
  <Form className=\\"todo-form\\">
    <Field id=\\"text\\" name=\\"text\\" type=\\"text\\" placeholder=\\"Add new Todo\\" disabled={false} innerRef={{...}} />
  </Form>
</Formik>"
`;
